juju/juju
# apiserver: support appdata in RemoteRelations and CrossModelRelations facades

### Checklist

   These aren't client facades, so no pylibjuju changes needed.
  Only API changes in this PR - the worker changes to change behaviour will follow.

----

## Description of change

The `CrossModelRelations.PublishRelationChanges` and `RemoteRelations.ConsumeRemoteRelationChanges` methods have been extended so they can apply updated application-level settings being sent with the incoming events (relying on the changes in #11004). 

The remote relations worker that pushes events to these methods reads from the other side by watching for changes in the relation. That was originally done using the same `RelationUnitsWatcher` that the uniter uses to watch for relation data changes within a model - the events it emits don't contain new settings, just notifications that some unit or application settings have changed or units have departed the relation. The client then has to call back to the controller to get the settings changes.

After some discussion with @jameinel and @wallyworld we realised that that the cross-model API would be a lot cleaner and more self-contained if it didn't require those extra round-trips to get the settings once the local or remote controller has notified the worker that some settings have changed. Those make sense in the uniter, which keeps track of settings versions it has seen and only goes back to the server for new versions when needed, but not in the remote relations worker which is essentially stateless and always needs to go back to get the settings before it can publish the change.

 To do this the two API versions have been bumped to v2: `CrossModelRelations.WatchRelationUnits` and RemoteRelations.WatchLocalRelationUnits have been removed, along with both of the `RelationUnitSettings` methods (which are used to get the individual settings). `CrossModelRelations.WatchRelationChanges` and `RemoteRelations.WatchLocalRelationChanges` have been added instead - these both return a new `RemoteRelationWatcher` which in turn emits `params.RemoteRelationChangeEvents` that can be passed on to the other side's `PublishRelationChanges` or `ConsumeRemoteRelationChanges` method.

A follow-up PR will update the client facades and remote relations worker to use the new API versions.

**WIP: still working on unit tests for these changes. Based on #11004 so includes the two commits for that PR.**

## QA steps

API only, no behaviour changes to use the new API version yet. Deploy a normal (non-appdata) cross-model relation and check that it still works.

```sh
juju deploy -m c1:source ~/juju/acceptancetests/repository/charms/dummy-source
juju deploy -m c2:sink ~/juju/acceptancetests/repository/charms/dummy-sink
juju offer -c c1 source.dummy-source:sink
juju relate -m c2:sink c1:source.dummy-source dummy-sink

juju config -m c1:source dummy-source token=testvalue
```
See that any sink application units see the updated token value.

## Documentation changes

None

## Bug reference

None
